File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.prefix_sum = [0] * len(nums)

        for i, num in enumerate(nums):
            if i == 0:
                self.prefix_sum[i] = nums[i]
            else:
                self.prefix_sum[i] = nums[i] + self.prefix_sum[i - 1]

    def sumRange(self, left: int, right: int) -> int:
            if left == 0:
                return self.prefix_sum[right]

            return self.prefix_sum[right] - self.prefix_sum[left - 1] 
        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */

var NumArray = function(nums) {
    let prev = 0;
    this.prefixSum = new Array(nums);
    
    for(let i=0; i<nums.length; i++) {
        this.prefixSum[i] = prev + nums[i];
        prev = this.prefixSum[i]
    }   
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
    if(left === 0)
        return this.prefixSum[right]
    
    return this.prefixSum[right] - this.prefixSum[left - 1]
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */